instance/models/mixins/database.py

```diff
@@ -225,25 +249,30 @@ def provision_mysql(self):
             self.mysql_provisioned = True
             self.save()
 
-    def deprovision_mysql(self):
+    def deprovision_mysql(self, ignore_errors=False):
         """
         Drop all MySQL databases and users.
         """
+        self.logger.info('Deprovisioning MySQL started.')
         if self.mysql_server and self.mysql_provisioned:
-            cursor = _get_mysql_cursor(self.mysql_server)
-
-            # Drop default databases and users
-            for database in self.mysql_databases:
-                database_name = database["name"]
-                _drop_database(cursor, database_name)
-                _drop_user(cursor, database["user"])
-
-            # Drop users with global privileges
-            for user in self.global_users:
-                _drop_user(cursor, user)
+            try:
+                cursor = _get_mysql_cursor(self.mysql_server)
+
+                # Drop default databases and users
+                for database in self.mysql_databases:
+                    database_name = database["name"]
+                    _drop_database(cursor, database_name)
+                    _drop_user(cursor, database["user"])
+                # Drop users with global privileges
+                for user in self.global_users:
+                    _drop_user(cursor, user)
+            except MySQLError:
+                if not ignore_errors:
+                    raise
 
             self.mysql_provisioned = False
             self.save()
+        self.logger.info('Deprovisioning MySQL finished.')
```

Comment:
Please see [this comment](https://github.com/open-craft/opencraft/pull/373#discussion_r241937262)
It would be much easier (and more reusable) to use something like this for logging:
```python
def action_logger(action: str):
    def decorator(f):
        def wrapper(self, *args, **kwargs):  
            self.logger.info('{} started.'.format(action)  
            f()
            self.logger.info('{} finished.'.format(action)  
        return wrapper
    return decorator
```

Reply:
Yes, this could be easily put into a decorator. But I think that would add an unnecessary layer of abstraction when looking at the logs.
The logs will tell us that the line where this log message comes from is not where the actual deprovision code is but where the decorator is implemented. To me, this is counterintuitive to what logs should be.
Although, if this is a standard we follow then I can move this to a decorator.

Reply:
> The logs will tell us that the line where this log message comes from is not where the actual deprovision code is but where the decorator is implemented. To me, this is counterintuitive to what logs should be.
Do you think it's important for info logs? We have the module name for them.

Reply:
Since the message is not general, as in "Deprovisioned started", but more specific, as in "Deprovisioning <resource> started/finished" I think it is important.
But it's not something that would break anything so I'm also OK to move it into a decorator.

Reply:
@xirdneh, I'm not sure what you're referring to. The example decorator above works like:

```python
@action_logger('Deprovisioning MySQL')
def deprovision_mysql(...):
    ...
```

So it's generic enough to handle all specific messages for our use cases, right? Am I missing something here?

Reply:
Yes, exactly. The log messages are not really generic because the actual message is passed as an argument into the decorator. If the message was generic then it wouldn't need to have a configurable action and could work for any deprovision method. So, really we're only saving a few characters an introducing something that is not clear what'll do right away (a decorator). Until you look at the decorator implementation that is.
Also, even if this is an info log message there's some context derived from the record, the module and line where the message is logged. If I were trying to get a sense about what is happening from the logs I would be directed to a function that is not where the actual deprovision is being implemented.

Now, these points of view are more preferences than anything and either way doesn't affect performance or clarity too much, at least not to my perception. If this is something that would align better with the project I'm up for it. What do you think, @Agrendalath ?

Reply:
Ok, let's leave it this way.